cmd/boostd/genindex.go

```diff
@@ -0,0 +1,221 @@
+			Usage:    "the path to the car",
+			Required: false,
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		ctx := lcli.ReqContext(cctx)
+
+		// parse piececid
+		piececid, err := cid.Decode(cctx.String("piececid"))
+		if err != nil {
+			return err
+		}
+		fmt.Println("piececid to index: ", piececid)
+
+		// connect to the piece directory service
+		pdClient := piecedirectory.NewStore()
+		defer pdClient.Close(ctx)
+		err = pdClient.Dial(ctx, cctx.String("api-piece-directory"))
+		if err != nil {
+			return fmt.Errorf("error while connecting to piece directory service: %w", err)
+		}
+
+		// connect to the full node
+		fnApiInfo := cctx.String("api-fullnode")
+		fullnodeApi, ncloser, err := lib.GetFullNodeApi(ctx, fnApiInfo, log)
+		if err != nil {
+			return fmt.Errorf("getting full node API: %w", err)
+		}
+		defer ncloser()
+
+		// connect to the storage API and create a sector accessor
+		storageApiInfo := cctx.String("api-storage")
+		sa, storageCloser, err := lib.CreateSectorAccessor(ctx, storageApiInfo, fullnodeApi, log)
+		if err != nil {
+			return err
+		}
+		defer storageCloser()
+
+		pr := &piecedirectory.SectorAccessorAsPieceReader{SectorAccessor: sa}
+
+		pd := piecedirectory.NewPieceDirectory(pdClient, pr, cctx.Int("add-index-throttle"))
+
+		var r *carv2.Reader
+		var headerDataSize uint64
+
+		if filepath := cctx.String("filepath"); filepath != "" {
+			fmt.Println("fetching piececid from car file: ", filepath)
+
+			r, err = carv2.OpenReader(filepath)
+			if err != nil {
+				return fmt.Errorf("open reader: %w", err)
+			}
+			defer r.Close()
+
+			if r.Version == 1 {
+				fi, err := os.Stat(filepath)
+				if err != nil {
+					return fmt.Errorf("os.stat err: %w", err)
+				}
+				headerDataSize = uint64(fi.Size())
+			}
+		} else {
+			fmt.Println("no car file specified, fetching piececid from the storage api")
+
+			sr, err := pd.GetPieceReader(ctx, piececid)
+			if err != nil {
+				if errors.Is(err, piecedirectory.ErrNoUnsealedPieceAvailable) {
+					return fmt.Errorf("piece reader could not find unsealed piece for cid %s: %w", piececid, err)
+				}
+				return fmt.Errorf("error while getting piece reader: %w", err)
+			}
+
+			r, err = carv2.NewReader(sr)
+			if err != nil {
+				return err
+			}
+		}
+
+		// gen index
+		recs, err := GetRecords(r, headerDataSize)
+		if err != nil {
+			return err
+		}
+
+		addStart := time.Now()
+
+		fmt.Printf("about to add %d records for piececid %s to the piece directory\n", len(recs), piececid)
+		err = pd.AddIndex(ctx, piececid, recs)
+		if err != nil {
+			return err
+		}
+
+		//TODO: maybe set car size?
+		//SetCarSize(ctx context.Context, pieceCid cid.Cid, size uint64) error
+
+		fmt.Println("adding index took", time.Since(addStart).String())
+
+		fmt.Printf("successfully added index (%d records) for piececid %s to the piece directory\n", len(recs), piececid)
+
+		return nil
+	},
+}
+
+func GetRecords(r *carv2.Reader, headerDataSize uint64) ([]model.Record, error) {
```

Comment:
I think instead of this function we can just use `PieceDirectory.buildIndexForPiece()`